enhancements/network/multi-network-vrfs.md

```diff
@@ -0,0 +1,924 @@
+this might be necessary, but for OpenShift purposes will only describe the behavior of IC in this proposal. In the layer
+2 IC model, GRs per node on a network will all be connected to the layer 2 transit switch:
+
+![Egress IP Layer 2](images/egress-ip-l2-primary.svg)
+
+In the above diagram, Node 2 is chosen to be the egress IP node for any pods in namespace A. Pod 1 and Pod 2 have
+default gateway routes to their respective GR on their node. When egress traffic leaves Pod 2, it is sent towards its
+GR-A on node 2, where it is SNAT’ed to the egress IP and the traffic sent to br-ex. For Pod 1, its traffic is sent to
+its GR-A on Node 1, where it is then rerouted towards GR-A on Node 2 for egress.
+
+##### Egress Firewall
+
+Egress firewall is enforced at the OVN logical switch, and this proposal has no effect on its functionality.
+
+##### Egress QoS
+
+Egress QoS is namespace scoped and functions by marking packets at the OVN logical switch, and this proposal has no
+effect on its functionality.
+
+##### Egress Service
+
+Egress service is namespace scoped and its primary function is to SNAT egress packets to a load balancer IP. As
+previously mentioned, the feature works the same in shared and local gateway mode, by leveraging the local gateway mode
+path. Therefore, its design will be covered in the Local Gateway Mode section of the Design Details.
+
+##### Multiple External Gateways (MEG)
+
+There will be no support for MEG or pod direct ingress on any network other than the primary, cluster default network.
+This support may be enhanced later by extending VRFs/networks outside the cluster.
+
+#### Local Gateway Mode
+
+With local gateway mode, egress/ingress traffic uses the kernel’s networking stack as a next hop. OVN-Kubernetes
+leverages an interface named “ovn-k8s-mp0” in order to facilitate sending traffic to and receiving traffic from the
+host. For egress traffic, the host routing table decides where to send the egress packet, and then the source IP is
+masqueraded to the node IP of the egress interface. For ingress traffic, the host routing table steers packets destined
+for pods via ovn-k8s-mp0 and SNAT’s the packet to the interface address.
+
+For multiple networks to use local gateway mode, some changes are necessary. The ovn-k8s-mp0 port is a logical port in
+the OVN topology tied to the cluster default network. There will need to be multiple ovn-k8s-mp0 ports created, one per
+network. Additionally, all of these ports cannot reside in the default VRF of the host network. Doing so would result in
+an inability to have overlapping subnets, as well as the host VRF would be capable of routing packets between namespace
+networks, which is undesirable. Therefore, each ovn-k8s-mp0-&lt;network> interface must be placed in its own VRF:
+
+![Local GW Node Setup](images/local-gw-node-setup-vrfs.svg)
+
+The VRFs will clone the default routing table, excluding routes that are created by OVN-Kubernetes for its networks.
```

Comment:
so you can think of it like this. Each user-defined network in OVN is a separate, and isolated network, regardless of its type (l2 or l3). Then this network is eventually routed out of the node. The routing of each of these networks in isolation, without knowing about one another is done in their own VRF instance. How this VRF instance manifests itself is in two ways:
1. As actual linux VRFs inside of the host
2. As VRFs using flows in br-ex

Since the external side of the node is not carrying these networks as VPNs across the external network, we have to leak these VRFs into the default network/VRF. In order to do that from a linux perspective, we need to leak routes from the default VRF into each VRF. For example, if on my host my default gw route to my next hop router is 0.0.0.0/0 via 192.168.1.1, that route would only appear in the default routing table, and not in my vrf's routing table. So when a packet from my user-defined network comes into my vrf and it wants to talk to the internet, it needs a route. Therefore we need to clone the route from the default routing table into the vrf so that it can get outside.

Using route leaking, we can allow pods in isolated networks to egress via default routing table, but restrict their vrf routes so that the isolated networks cannot communicate with each other. Also, thanks to conntrack we can track these packets so we know on reply packets which network these packets should be sent to.